Add support for dedicated nodefs-imagefs storage location

shimpa1 opened the pull request:
If this PR is accepted, all providers built with these scripts will have their respective imagefs/nodefs directory located in /ephemeral, regardless if a separate storage is being used for this purpose or not. The original /var/lib/containerd and /var/lib/kubelet directories will no longer be used.

If a new provider has dedicated imagefs/nodefs storage, as most high-end providers do, make sure the storage is mounted on /ephemeral in r/w mode.

shimpa1 committed:
Update workerNode.sh

added support for custom nodefs/imagefs location

shimpa1 committed:
Update k3sAndProviderServices.sh

added support for custom imagefs/nodefs location

shimpa1 committed:
Update README.md

shimpa1 committed:
Update README.md

shimpa1 committed:
Update k3sAndProviderServices.sh

shimpa1 committed:
Update workerNode.sh

shimpa1 committed:
Update workerNode.sh

shimpa1 committed:
Update k3sAndProviderServices.sh

shimpa1 committed:
Update k3sAndProviderServices.sh

shimpa1 committed:
Update providerBuild.sh

andy108369 reviewed (changes requested):
Great start @shimpa1 
After having a discussion, please:
- make the custom data path optional (configurable) to not break existing provider configs;
- add the example of how to build a cluster using the new custom data path argument;
- worth adding some note for Ceph config, since Ceph requires this to be set when custom data path is used, e.g.:
  - see `kubeletDirPath` under `Make applications aware of the new nodefs location` in https://akash.network/docs/providers/build-a-cloud-provider/kubernetes-cluster-for-akash-providers/kubernetes-cluster-for-akash-providers/#make-applications-aware-of-the-new-nodefs-location

shimpa1 commented:
thanks @andy108369 on it

shimpa1 committed:
Update k3sAndProviderServices.sh

shimpa1 committed:
Update workerNode.sh

shimpa1 committed:
Update workerNode.sh

shimpa1 committed:
Update README.md

shimpa1 committed:
Update README.md

shimpa1 committed:
Update README.md

shimpa1 committed:
Update README.md

andy108369 commented on providerBuild.sh:
```diff
@@ -253,9 +253,13 @@ if [ "$install_storage_support" = true ]; then
     # Install the Rook-Ceph Helm chart for the operator
     echo "Installing Rook-Ceph operator..."
-    helm install --create-namespace -n rook-ceph rook-ceph rook-release/rook-ceph --version 1.14.0
+    helm install --create-namespace -n rook-ceph rook-ceph rook-release/rook-ceph --version 1.14.0 -f rook-ceph-operator.values.yml
```
let's bump rook-ceph from `1.14.0` to `1.15.6` ([ref](https://github.com/rook/rook/releases/tag/v1.15.6))

andy108369 commented on providerBuild.sh:
```diff
@@ -253,9 +253,13 @@ if [ "$install_storage_support" = true ]; then
     echo "Rook-Ceph repository added."
 
+    # Create the rook-ceph-operator.values.yaml file and append the kubelet directory location
+    touch /root/provider/rook-ceph-operator.values.yaml
+    echo "kubeletDirPath: /ephemeral/kubelet" >> /root/provider/rook-ceph-operator.values.yaml
```
no need for `touch` here, since `>` (overwrite) or `>>` (append) will create the file anyway if it does not exist.

Also, if `providerBuild.sh` runs more than once - it will keep appending (`>>`) the `kubeletDirPath: /ephemeral/kubelet` line in `/root/provider/rook-ceph-operator.values.yaml` file.

Is providerBuild.sh supposed to be ran more than once? If so, then there should be a check if this line already present in the yaml and skip appending it; otherwise the yaml will have multiple of such lines.

shimpa1 replied:
The providerBuild.sh script is supposed to be run once.

andy108369 commented on k3sAndProviderServices.sh:
```diff
@@ -56,6 +58,12 @@ while getopts ":d:e:tagm:c:r:w:n:s:" opt; do
       exit 1
       ;;
+    k )
+      kubelet_storage="--kubelet-arg=root-dir=$OPTARG"
```
We can pass not only the `root-dir` argument using `--kubelet-arg`, but also other options if needed. 

We could address this either in the current PR or in a future update, should we need to configure additional kubelet options via `--kubelet-arg` beyond just `root-dir`.

How this could be addressed is:

1. move line 18 after line 73 (after `shift $((OPTIND -1))`) and extend it with `${kubelet_args}`
```
kubelet_args=""
k3s_common_args="--disable=${disable_components} --flannel-backend=none --kubelet-arg=${kubelet_args}"
```

2. for `k )` argument, use newly introduced `kubelet_args`
```
    k )
      kubelet_args="--root-dir=$OPTARG"
```

3. Remove `$kubelet_storage` from line 185 since `kubelet_args` with `--root-dir` will come via `k3s_common_args`, e.g.:
```
    curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="$k3s_common_args $install_exec $containerd_storage" sh -
```

And leave `$containerd_storage` as-is.

andy108369 replied:
@shimpa1  since it requires more work and we need to build new providers rather ASAP now -- we can tackle this comment later in a separate PR (should then create a new issue from [this](https://github.com/chainzero/provider-build-scripts/pull/19/files#r1877892003) comment).

shimpa1 replied:
I agree.
Both K3S build scripts can be used as they are right now and we can update them later.

shimpa1 committed:
Update providerBuild.sh

shimpa1 committed:
Update README.md

shimpa1 committed:
Update providerBuild.sh

shimpa1 committed:
Update providerBuild.sh

shimpa1 committed:
Update README.md

shimpa1 committed:
Update providerBuild.sh

shimpa1 commented on providerBuild.sh:
@andy108369 
Please review providerBuild.sh and comment!

ghost commented:
Will use https://github.com/chainzero/provider-build-scripts/pull/19/commits/1cf78ab977ab7268f774144da45e91de369347d8 version

shimpa1 committed:
Update providerBuild.sh

added IMAGEFS_DIR detection for future use
fixed path extraction from the service file, it no longer needs to be trimmed later in the script

shimpa1 committed:
Update workerNode.sh

andy108369 commented on README.md:
```diff
@@ -1,10 +1,8 @@
-`mkdir -p /ephemeral`
-
-If the cluster has a dedicated nodefs/imagefs storage (a separate nvme drive or a number of drives in a RAID) make sure it's mounted on `/ephemeral`
+If the cluster has a dedicated ephemeral storage, specify the location using -o and -k
```
let's keep the `nodefs`/`imagefs` bits for clarity, e.g.:
```
If the cluster has a dedicated storage for nodefs/imagefs (aka ephemeral K8s data), specify the location using -o and -k
```

andy108369 commented on providerBuild.sh:
```diff
@@ -3,18 +3,55 @@
+
+# Look for root-dir (nodefs) argument
+KUBELET_ARG=$(echo "$EXECSTART" | grep -o "\--kubelet-arg=root-dir=[^ ]*")
+NODEFS_DIR=$((extract_path "$KUBELET_ARG" "$DEFAULT_KUBELET_DIR") | tr -d "'")
```
I guess this should be the IMAGEFS_DIR (derived from --kubelet-arg=root-dir=)

andy108369 commented on providerBuild.sh:
```diff
@@ -3,18 +3,55 @@
+KUBELET_ARG=$(echo "$EXECSTART" | grep -o "\--kubelet-arg=root-dir=[^ ]*")
+NODEFS_DIR=$((extract_path "$KUBELET_ARG" "$DEFAULT_KUBELET_DIR") | tr -d "'")
+IMAGEFS_ARG=$(echo "$EXECSTART" | grep -o "\--data-dir=[^ ]*")
+IMAGEFS_DIR=$((extract_path "$IMAGEFS_ARG" "$DEFAULT_IMAGEFS_DIR") | tr -d "'")
```
And this should be the NODEFS_DIR (derived from the --data-dir)

andy108369 commented on providerBuild.sh:
```diff
@@ -3,18 +3,55 @@
+### LOGIC THAT DETERMINES EPHEMERAL STORAGE LOCATION
+
+# Default paths
+DEFAULT_KUBELET_DIR="/var/lib/kubelet"
```
for clarity - this is the default NODEFS directory (valid for both K3S & Kubespray based deployments)

andy108369 commented on providerBuild.sh:
```diff
@@ -3,18 +3,55 @@
+
+# Default paths
+DEFAULT_KUBELET_DIR="/var/lib/kubelet"
+DEFAULT_CONTAINERD_DIR="/var/lib/containerd"
```
This is the IMAGEFS directory.
Which is /var/lib/containerd for kubespray based deployments.
/var/lib/rancher/k3s/agent/containerd - for K3S based deployments.

andy108369 commented on workerNode.sh:
```diff
@@ -26,6 +29,12 @@ while getopts ":m:t:gs" opt; do
+      kubelet_storage="--kubelet-arg=root-dir=$OPTARG"
+      ;;
+    o )
+      containerd_storage="--data-dir=$OPTARG"
```
Let's name it nodefs_dir

andy108369 commented on workerNode.sh:
```diff
@@ -26,6 +29,12 @@ while getopts ":m:t:gs" opt; do
       install_storage_support=true
       ;;
+    k )
+      kubelet_storage="--kubelet-arg=root-dir=$OPTARG"
```
Let's name it imagefs_dir

shimpa1 committed:
Update providerBuild.sh

shimpa1 commented:
closing, cleaning, will reopen
